pkg/servicemesh/control-plane.go

```diff
@@ -0,0 +1,414 @@
+import (
+	"context"
+	"fmt"
+
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	"github.com/openshift-kni/eco-goinfra/pkg/msg"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	istioV2 "maistra.io/api/core/v2"
+	goclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// ControlPlaneBuilder provides a struct for serviceMeshControlPlane object from the cluster and
+// a serviceMeshControlPlane definition.
+type ControlPlaneBuilder struct {
+	// serviceMeshControlPlane definition, used to create the serviceMeshControlPlane object.
+	Definition *istioV2.ServiceMeshControlPlane
+	// Created serviceMeshControlPlane object.
+	Object *istioV2.ServiceMeshControlPlane
+	// Used in functions that define or mutate serviceMeshControlPlane definition. errorMsg is processed
+	// before the serviceMeshControlPlane object is created
+	errorMsg string
+	// api client to interact with the cluster.
+	apiClient *clients.Settings
+}
+
+// NewControlPlaneBuilder method creates new instance of builder.
+func NewControlPlaneBuilder(apiClient *clients.Settings, name, nsname string) *ControlPlaneBuilder {
+	glog.V(100).Infof("Initializing new serviceMeshControlPlane ControlPlaneBuilder structure with the following "+
+		"params: name: %s, namespace: %s", name, nsname)
+
+	builder := &ControlPlaneBuilder{
+		apiClient: apiClient,
+		Definition: &istioV2.ServiceMeshControlPlane{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      name,
+				Namespace: nsname,
+			},
+		},
+	}
+
+	if name == "" {
+		glog.V(100).Infof("The name of the serviceMeshControlPlane is empty")
+
+		builder.errorMsg = "The serviceMeshControlPlane 'name' cannot be empty"
+	}
+
+	if nsname == "" {
+		glog.V(100).Infof("The namespace of the serviceMeshControlPlane is empty")
+
+		builder.errorMsg = "The serviceMeshControlPlane 'namespace' cannot be empty"
+	}
+
+	return builder
+}
+
+// WithAllAddonsDisabled disables all addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithAllAddonsDisabled() *ControlPlaneBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	enablement := false
+
+	glog.V(100).Infof(
+		"Creating serviceMeshControlPlane %s with the all addons disabled", builder.Definition.Name)
+
+	disableAllAddons := istioV2.AddonsConfig{
+		Prometheus: &istioV2.PrometheusAddonConfig{
+			Enablement: istioV2.Enablement{Enabled: &enablement},
+		},
+		Grafana: &istioV2.GrafanaAddonConfig{
+			Enablement: istioV2.Enablement{Enabled: &enablement},
+		},
+		Kiali: &istioV2.KialiAddonConfig{
+			Enablement: istioV2.Enablement{Enabled: &enablement},
+		},
+		ThreeScale: &istioV2.ThreeScaleAddonConfig{
+			Enablement: istioV2.Enablement{Enabled: &enablement},
+		},
+	}
+
+	builder.Definition.Spec.Addons = &disableAllAddons
+
+	return builder
+}
+
+// WithGrafanaAddon adds grafana addons to the serviceMeshControlPlane.
+func (builder *ControlPlaneBuilder) WithGrafanaAddon(
```

Comment:
Please add validation for all parameters. Check my comment: https://github.com/openshift-kni/eco-goinfra/pull/147/files#r1362669123